knora/dsplib/utils/xml_upload.py

```diff
@@ -910,94 +883,435 @@ def xml_upload(input_file: str, server: str, user: str, password: str, imgdir: s
+                }
+                jsondata = json.dumps(jsonobj, indent=4, separators=(',', ': '))
+                response = try_network_action(
+                    object=con,
+                    method='post',
+                    args={'path': '/v2/values', 'jsondata': jsondata},
+                    terminal_output_on_failure=f'ERROR while uploading the resptr prop of "{link_prop.name}" '
+                                               f'of resource "{resource.id}"'
+                )
+                if not response:
+                    continue
+                stashed_resptr_props[resource][link_prop].remove(resptr)
+                if verbose:
+                    print(f'  Successfully uploaded resptr-prop of "{link_prop.name}"\n'
+                          f'    Value: {resptr}')
+
+    # make a purged version of stashed_resptr_props, without empty entries
+    nonapplied_resptr_props = purge_stashed_resptr_props(stashed_resptr_props)
+    return nonapplied_resptr_props
+
+
+def try_network_action(
+    terminal_output_on_failure: str,
+    method: Union[str, Callable[..., Any]],
+    object: Optional[Any] = None,
+    args: Optional[dict[str, Any]] = None
+) -> Any:
+    """
+    Helper method that tries 7 times to execute an action. Each time, it catches ConnectionError and
+    requests.exceptions.RequestException, which lead to a waiting time and a retry. The waiting times are 1,
+    2, 4, 8, 16, 32, 64 seconds. It also catches BaseError and Exception each time, which lead to a message being
+    printed and None being returned.
+    If there is still no success at the end, the message is printed and None is returned.
+
+    Args:
+        terminal_output_on_failure: message to be printed if action cannot be executed
+        method: either a callable to be called on its own, or a method name (as string) to be called on object
+        object: if provided, it must be a python variable/object, accompanied by a method name (as string)
+        args: if provided, a dict with the arguments passed to method
+
+    Returns:
+        the return value of action, or None
+    """
+
+    for i in range(7):
         try:
-            update_resptr_props(
-                resource=resource,
-                res_iri=res_iri,
-                prop_2_resptrs=prop_2_resptrs,
-                res_iri_lookup=res_iri_lookup,
-                con=con,
-                verbose=verbose
-            )
-        except BaseError as err:
-            print(f'BaseError while updating an XML text of resource "{resource.id}": {err.message}')
+            if object and isinstance(method, str):
+                if not args:
+                    return getattr(object, method)()
+                else:
+                    return getattr(object, method)(**args)
+            else:
+                if not args:
+                    return method()
+                else:
+                    return method(**args)
```

Comment:
see my comment above regarding this: if it works to pass a class method as a function to this method, then you wouldn't need the `object` and distinction between string and callable, so this if-else could be removed.

And I think you can pass an empty dictionary as kwargs, so if you'd simply default it to `{}`, you could get rid of the inner if-else too.